activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb

```diff
@@ -191,10 +195,18 @@ def []=(sql, key)
 
         private
           def dealloc(key)
-            @connection.query "DEALLOCATE #{key}" if connection_active?
+            @connection.query "DEALLOCATE #{key}" if connection_active? && has_prepared_statement?(key)
           rescue PG::Error
           end
 
+          def has_prepared_statement?(key)
+            result = @connection.query "SELECT COUNT(name) AS key_count FROM pg_prepared_statements WHERE name = '#{key}'"
```

Comment:
Would not querying the prepared statement every time you want to create a new one be slower? @sgrif suggestion was to try to deallocate without having to check if it exists or not

Reply:
@rafaelfranca  Yes, it would if it was without penalty. However, deallocating statement names that don't exist throws an exception. This is likely fine, if you are not inside a transaction, but if you are, then that transaction will fail going forward with:

`ActiveRecord::StatementInvalid: PG::InFailedSqlTransaction: ERROR:  current transaction is aborted, commands ignored until end of transaction block`

Reply:
What if we rescue the exception and just return `nil` in the same way we do with `PG::Error`?

Reply:
@rafaelfranca Have resolved this per your and @matthewd suggestions.